tests/msc2403_test.go

```diff
@@ -0,0 +1,241 @@
+		query := url.Values{
+			"server_name": []string{"hs1"},
+		}
+
+		knockingUser.MustDoWithStatusRaw(
+			t,
+			"POST",
+			[]string{"_matrix", "client", "r0", "join", roomID},
+			[]byte("{}"),
+			"application/json",
+			query,
+			403,
+		)
+	})
+
+	t.Run("Knocking on a room with join rule 'xyz.amorgan.xyz' should succeed", func(t *testing.T) {
+		knockOnRoom(t, knockingUser, roomID, testKnockReason, []string{"hs1"})
+	})
+
+	t.Run("A user that has already knocked cannot immediately knock again on the same room", func(t *testing.T) {
+		knockOnRoomWithStatus(t, knockingUser, roomID, "I really like knock knock jokes", []string{"hs1"}, 403)
+	})
+
+	// Now that Bob's knocked on the room successfully, let's check that everything went right
+
+	t.Run("parallel", func(t *testing.T) {
+		t.Run("Users in the room see a user's membership update when they knock", func(t *testing.T) {
+			t.Parallel()
+			inRoomUser.SyncUntilTimelineHas(t, roomID, func(ev gjson.Result) bool {
+				if ev.Get("type").Str != "m.room.member" || ev.Get("sender").Str != knockingUserUserID {
+					return false
+				}
+				must.EqualStr(t, ev.Get("content").Get("reason").Str, testKnockReason, "incorrect reason for knock")
+				must.EqualStr(t, ev.Get("content").Get("membership").Str, "xyz.amorgan.knock", "incorrect membership for knocking user")
+				return true
+			})
+		})
+
+		t.Run("Users see state events from the room that they knocked on in /sync", func(t *testing.T) {
+			t.Parallel()
+			knockingUser.SyncUntil(
+				t,
+				"",
+				"rooms."+client.GjsonEscape("xyz.amorgan.knock")+"."+client.GjsonEscape(roomID)+".knock_state.events",
+				func(ev gjson.Result) bool {
+					// We don't current define any required state event types to be sent
+					// If we've reached this point, then an entry for this room was found
+					return true
+				},
+			)
+		})
+	})
+
+	if !federation {
+		// Rescinding a knock over federation is currently not supported in Synapse
+		//
+		t.Run("A user that has knocked on a local room can rescind their knock and then knock again", func(t *testing.T) {
+			// Rescind knock
+			knockingUser.MustDo(
+				t,
+				"POST",
+				[]string{"_matrix", "client", "r0", "rooms", roomID, "leave"},
+				struct {
+					Reason string `json:"reason"`
+				}{
+					"Just kidding!",
+				},
+			)
+
+			// Knock again
+			knockOnRoom(t, knockingUser, roomID, "Let me in... again?", []string{"hs1"})
+		})
+	}
+
+	t.Run("A user in the room can reject a knock", func(t *testing.T) {
+		// Reject the knock
+		inRoomUser.MustDo(
+			t,
+			"POST",
+			[]string{"_matrix", "client", "r0", "rooms", roomID, "kick"},
+			struct {
+				UserID string `json:"user_id"`
+				Reason string `json:"reason"`
+			}{
+				knockingUserUserID,
+				"I don't think so",
+			},
+		)
+
+		// Knock again
+		knockOnRoom(t, knockingUser, roomID, "Pleeease let me in?", []string{"hs1"})
+	})
+
+	t.Run("A user in the room can accept a knock", func(t *testing.T) {
```

Comment:
I think a knock after this isn't allowed and should 403?

Reply:
True! I'll stick in a test for that, thanks!